### PR TITLE
Sync channel membership after connecting room

### PR DESCRIPTION
**Pull request by Half-Shot**
It's a fairly each change to make the experience better :)

As @Cadair requested, this now syncs joins and leaves incrementally as well as ensures they sync on startup.

### CODE REVIEW

File: src/Main.ts [hunk 1]
```diff
@@ -1114,6 +1114,11 @@ export class Main {
             }
             room.setBotClient(slackClient);
             room.SlackChannelName = infoRes.channel.name;
+
+            // Perform syncing asyncronously.
```
Comment (jaller94) [inline src/Main.ts:1118]:
```suggestion
            // Perform syncing asynchronously.
```

File: src/Main.ts [hunk 2]
```diff
@@ -1114,6 +1114,11 @@ export class Main {
...
+            this.teamSyncer?.syncMembershipForRoom(matrixRoomId, opts.slack_channel_id, opts.team_id, slackClient).catch((ex) => {
```
Comment (jaller94) [inline src/Main.ts:1119]: Do you not want to await the sync, so it doesn't race with the code below?

**Comment by Cadair** [comment]
What about also doing this for all existing channels? :angel:

**Comment by Half-Shot** [comment]
>What about also doing this for all existing channels? angel

This should be done on startup. `TeamSyncer.syncAllTeams` is called, which calls `syncItems(teamId, client, "channel");` which calls `syncChannel`. We also sync membership events from slack as they happen.

File: src/Main.ts (hunk 2 above)
Comment (Half-Shot) [reply src/Main.ts]: Moved this to the end.

File: src/TeamSyncer.ts [hunk 3]
```diff
@@ -208,11 +210,22 @@ export class TeamSyncer {
...
-    public async syncUser(teamId: string, domain: string, item: ISlackUser) {
+    public async syncUser(teamId: string, domain: string, item: ISlackUser, newUser = false) {
         log.info(`Syncing user ${teamId} ${item.id}`);
         const slackGhost = await this.main.ghostStore.get(item.id, domain, teamId);
         if (item.deleted !== true) {
             await slackGhost.updateFromISlackUser(item);
+            for (const teamRoom of this.main.rooms.getBySlackTeamId(teamId)) {
+                if (teamRoom.IsPrivate || teamRoom.SlackType !== "channel") {
+                    // We only want PUBLIC rooms
+                    return;
```
Comment (jaller94) [inline src/TeamSyncer.ts:221]: I think you meant to skip of this one room, but not terminate the sync for the user.
```suggestion
                    continue;
```

**Pull request merged**